CMT merge

RonRahaman opened the pull request:
This is the first update to the CMT code since November.  It includes many improvements to particle tracking, among other things.  

This branch results from two merges and a rebase between the UFCCMT BitBucket repo and the Nek5000 GitHub repo.  Let these repos be called 'UFCCMT' and 'Nek5000' respectively. Our workflow was the following:
- The directories of UFCCMT was restrucuted using 'git filter-branch'.  This was necessary because UFCCMT had the directory structure of the SVN repo.  
- The first merge was performed:
  1. We identified the last time UFCCMT had been synced with Nek5000.  That sync was a manual copy-paste of source files, not a Git merge.  
  2. A new branch (UFCCMT:cmt-sync) was created on UFCCMT:master at the downstream commit identified in (1).  
  3. A new branch (Nek5000:cmt-merge-1) was created on Nek5000:develop at the downstream commit identified in (1).  
  4. We merged UFCCMT:cmt-sync into Nek5000:cmt-merge-1
- A rebase was performed:
  1. A new branch (UFCCMT:cmt-rebase-1) was created at the most recent commit of UFCCMT:master.
  2. UFCCMT:cmt-rebase-1 was rebased on top of Nek5000:cmt-merge-1
  3. The rebased UFCCMT:cmt-rebase-1 was pushed to Nek5000 as a new branch (Nek5000:cmt-rebase-1)
- A second and final rebase was performed:
  1. A new branch (Nek5000:cmt-merge-2) was created at the most recent commit of Nek5000:develop.
  2. Nek5000:cmt-rebase-1 was merged into Nek5000:cmt-merge-2

Rodbourn committed:
Initial commit from CMT-Nek CVS repo

Rodbourn committed:
Added title to readme file

Rodbourn committed:
Added a sub heading of team to demonstrate a pull request.

Rodbourn committed:
Merged in rodbourn/cmt-nek (pull request #1)

Added a sub heading of team to demonstrate a pull request.

hackljf committed:
Merged ccmt/cmt-nek into master

Rodbourn committed:
Revert "Added a sub heading of team to demonstrate a pull request."

This reverts commit e8af2372eb71c01a86cd8e371f53198ce195d9fd.

Commit:
Test commit option in SourceTree

Commit:
Created cmt_examples folder. Added first example, inviscid vortex. This case has been verified with the current version of CMT-nek (mrugeshs fork).

Commit:
Rarefaction example added to cmt_examples. Verified using current version of CMT-nek (mrugeshs fork)

Commit:
Modifications
drive2_cmt.f -  we do not initialize cmt parameters from setup_cmt_param. We initialize them in usrdat or usrdat2 in <casename>.usr. This is done to avoid trampling over param array declared in INPUT in nek5000 codebase.
drive1_cmt.f - move compute_primitive_vars is now called after the 3 stages of RK3 step are done. This ensures that we have access to the latest primitive variables are available in userchk.
diagnostics.f - out_pvar_nek and out_fld_var are modified. We use scrns common block to copy conserved variables and primitive variables and then passed as arguments to outpost2

Commit:
Merged mrugeshs/cmt-nek into master
- Added cat_examples
- Modifications to drive1_cmt.f, drive2_cmt.f and diagnostics.f

Commit:
Merged ccmt/cmt-nek into master

Rodbourn committed:
Merged ccmt/cmt-nek into master

Commit:
The inviscid vortex case has horizontal boundaries as walls. This modification makes them periodic. This case is meant to have periodic boundary condition in x and y direction.

Commit:
Swept_particles example added.

Commit:
Readme for swept particles

Commit:
Merged in mrugeshs/cmt-nek (pull request #2)

Master

Commit:
Merged ccmt/cmt-nek into master

hackljf committed:
Merged ccmt/cmt-nek into master

Commit:
Following is the summary of the changes
1) toteq has been moved to SIZE
   -  makenek.inc will search for toteq, if not found, declares toteq=1 when IFCMT=.false. or 5 when IFCMT=.true.
      NOTE that if the user uses a SIZE from a cmt-nek case to run a nek5000 example then makenek will not modify toteq !
2) CMTSIZE is removed from examples, all the parameters in CMTSIZE are declared in CMTDATA
3) CMT trunk has been modified. LELT and LELV have been replaced by LELCMT.
    This modification ensures that when the user is running nek5000 then I can set LELCMT=1 and the memory footprint of cmt-nek
    will be minimized. LELCMT is set by makenek.inc. The procedure is same as that used to set toteq.
4) cmt directory now resides inside nek
5) Fixed an error in makefile.template, CMT_CORE is always included in the list of object files.

Commit:
Merged in mrugeshs/cmt-nek (pull request #3)

Following is a summary of the changes

Rodbourn committed:
Merged ccmt/cmt-nek into master

Rodbourn committed:
Added tools from nek5000 into a tools folder since they are used in the examples (prevent having to fetch code from nek5000 repo)

Rodbourn committed:
Merged in rodbourn/cmt-nek (pull request #4)

Added tools from nek5000 into a tools folder since they are used in the examples (prevent having to fetch code from nek5000 repo)

Rodbourn committed:
Added a cmt_cases directory to contain cases which are under development.

Rodbourn committed:
Added ignore rules to remove some of the build and run artifacts

Rodbourn committed:
Added a quick description of how to run an example

Rodbourn committed:
Added further ignore patterns.

Rodbourn committed:
Merged in rodbourn/cmt-nek (pull request #5)

Master

Commit:
Merged ccmt/cmt-nek into master

Rodbourn committed:
Merged ccmt/cmt-nek into master

hackljf committed:
Merged ccmt/cmt-nek into master

Rodbourn committed:
Added cmt_cases/charles_particles to contain further particle development work which is initially a copy of the swep_particles example.  cmt_cases is meant to be cases in development rather than complete examples.

Rodbourn committed:
fix variable replacement which occurred during review.  Added map and rea

Rodbourn committed:
Added a quick build script that will copy makenek if needed.

Rodbourn committed:
Merge branch 'master' of bitbucket.org:rodbourn/cmt-nek

Rodbourn committed:
Merged in rodbourn/cmt-nek (pull request #6)

Started charles_particles case from swept particles

hackljf committed:
Merged ccmt/cmt-nek into master

Commit:
Merging nek5_svn commit of November 4th 2015.
appended few more excursions to the gitignore file

Commit:
this repo now resembles the folder hierarchy of nek_svn. CMT examples have been moved inside examples folder. To access cmt examples look in examples/CMT/
All the nek5000 examples are also added to this repo for the convenience of developers and users at CCMT.

Commit:
Merged in mrugeshs/cmt-nek (pull request #7)

Master

hackljf committed:
Merged ccmt/cmt-nek into master

hackljf committed:
Extensive changes for weak-form, dealiased integration. This is almost
certainly the right way to do nonlinear DG in nek5000 with Patera elements.
Extensive testing must precede an SVN commit, or even a push to
bitbucket.org/ccmt/master

hackljf committed:
boundary condition repair. thanks, Mrugesh

hackljf committed:
cleaned up some comments

hackljf committed:
Merged in hackljf/cmt-nek (pull request #8)

Master

hackljf committed:
Merged ccmt/cmt-nek into master

hackljf committed:
added case sensitivity to 'v  ' in InviscidFlux.
Added a kludge to inflow_bc until I can get cbu nekasgn'ed correctly.

hackljf committed:
Merge branch 'master' of https://bitbucket.org/hackljf/cmt-nek
ok apparently master wins

hackljf committed:
updated rarefaction.usr; need new grid soon

hackljf committed:
README.md edited online with Bitbucket

Commit:
MS021816 Update Jason's Fork with cmt_dummy file
This is to check if I can use git the way it is ment to be.

 Changes to be committed:
	new file:   nek/cmt/cmt_dummy.f

Commit:
Fix for restart cmt-nek and modifications to makenek.inc
and makefile.template to incorporate cmt_dummy option for compilation

Commit:
Merge branch 'master' of https://bitbucket.org/hackljf/cmt-nek

Commit:
Merged in hackljf/cmt-nek (pull request #9)

Master

Commit:
MS022416 Make dealiasing optional.
         If LXD=LX1 in size file then dealiasing is turned off
         in cmt-nek. Limited tests performed on inviscid
         compressible vortes and rarefaction case.

Commit:
my git example

dpzwick committed:
Basic giut tutorial

Commit:
Merge branch 'master' of https://bitbucket.org/ccmt/cmt-nek

dpzwick committed:
Revert "MS022416 Make dealiasing optional."

This reverts commit ca04b798f1c11d18f5a6ea2ed22089ebb28b1797.

dpzwick committed:
Current developement cases for particles and forces

dpzwick committed:
Working on particle forces and organizing cmt_cases folder

dpzwick committed:
Fixed bug in mpi io and added quasi-steady force corrections in the .usrp file rarepart2d case. Still in testing.

dpzwick committed:
Rearranged david_* cases from cmt_cases to examples/CMT/ directory

dpzwick committed:
synced folders from previous commit

Commit:
Adding sphere cases to CMT examples. These cases have an inflow outflow
 boundary. Two cases are added, one is meant to get a steady state solution
 and the other case is used to extract the inviscid unsteady kernel

dpzwick committed:
Added inout2d example case, but still a work in progress. Used for testing quasi-steady force corrections at different mach/reynolds numbers

dpzwick committed:
updated david_inout2d case for particle testing to fix bug in include file

Commit:
Adding optional dealiasing in cmt-nek
The switch is controlled by setting lxd in SIZE file,
if lxd.eq.lx1 then dealiasing is turned off

dpzwick committed:
Updated particle tracking to include RK3 time step routine so that there were minimal changes
in the trunk code. Quasi-steady force corrections have been added according to parmar et al. 2010.
Now the setup to add particles is as follows:

    1) Add the file cmtparticles.usrp to your directory
    2) Add the line
        include 'cmtparticles.usrp'
       to the begining of the reuglar .usr file
    3) In the following three lines to the userchk file
        include 'CMTDATA'
        inquire(file='cmtparticles.usrp',exist=arethereparticles)
        if (arethereparticles) call stokes_particles

       Note that the last two lines need to be added only in the loop
       where istep == 0. This step initializes the parameters for particles

    4) Change the starting locations, number of particles, and particle
       properites to suit the problem geometry and setup

    Please note that the files changed in the trunk directory were only CMTDATA
    and drive1_cmt.f.

    Along with this, examples have been added to examples/CMT/david_*:
       1) david_inout2d:
            super or subsonic inflow and outflow in a long domain
            so that particles have time to adjust to flow. Set mach number
            in two places in the .usr file to change the flow mach number.
       2) david_rarepart2d:
            rarefaction with particles in 2d
       3) david_rarepart3d:
            rarefaction with particles in 3d
       4) david_swept_particles
            same as the svn case except updated to include the newest particle udpates

dpzwick committed:
Updated test cases so that particles are consistent with drive1_cmt.f and
CMTDATA files. Also updated these files too so that two way coupling and
adding other forces should be straight forward.

dpzwick committed:
Updated and fixed some weird behavior with the low storage rk3 time stepping routine
for particles.

Moved cases from david_* to examples/rk3/*.

Particle velocity is now output in binary files velxyz*.3D now.

"partview" bash script has been made more general so it can handle velocity binary file conversion.
See the partview files in the rk3 directory to see how to call it. It can now be called with different
filenames and also 2d and 3d (2 or 3 is the last input argument into script).

Updated makenek.inc so that particle routines are added to the .usr file if cmtparticles.usrp is not in
the directory. Note that if you want particles, you still must add line at top of .usr file: include "cmtparticles.usrp"
even if the file is in the directory.

Updated CMTDATA file so that correct TVD Rk3 arrays are transfered correctly between timesteps.

dpzwick committed:
Updated the case rarepart2d and the cmtparticles.usrp file found within.

Added undisturbed force and inviscid unsteady force (both have no finite
mach/reynolds number corrections). Also updated the partview file. See
comments in the file for directions on how to call it. Also, there are
partdata* files that are output every output call that simily have ldim
and number of particles. This file must be copied to wherever partview is run
at. This is due to an outflow boundary condition allowed for particles where
they are destroyed if they are outside the domain (see the subroutine
update_particle_location(bc_part) in the .usrp file). Note that bc_part = 1
means periodic particles and bc_part = 0 means outflow where particles are destroyed.

Commit:
Sync up with latest nek svn repo.

Commit:
MS-04/18/16 Sync with nek5000 svn code Rev 1131. Paul has fixed
  maxjob and restart errors/bugs pointed out by cmt-nek micro sim.

dpzwick committed:
Added particle ability to restart. In order to restart (pending fluid
restart issues), set:
    1) IFHEAT to T in .rea file
    2) PRESOLVE line in .rea file from 0 to 1, also set fld file
       that you want to restart from on line below (exactly the same
       as nek5000)
    3) see included lines in example part_*/.usr files in usrdat2, but
       you must set ifrestart-.true. and also set the number of the particle
       file that you're restarting from (should be 1 less than fld restart file
       number).

hackljf committed:
JH050316 ic.f fixed to allow restarting (again-ish).

hackljf committed:
initialization is getting more, not less screwed up; find an answer

hackljf committed:
absurdly huge bugfix in compute_transport_props

RonRahaman committed:
Merge branch 'cmt-sync' into cmt-merge

This is the first merge in the big CMT Git sync (began 07/22/2016).  We
have merged the CMT repo bitbucket repo (from 04/18/2016) onto the
develop branch of the Nek5000 Git repo (from 04/20/2016).  The
core/cmt/*.f and makefile.template files from the CMT-bitbucket repo are
the only modified files.  The remaining files have been kept from the
Git repo without modification, as they are more recent.

RonRahaman committed:
Merge branch 'cmt-rebase-1' into cmt-merge-2

This is the second (final?!) merge of the big CMT sync.  Previously:
1. We identified the last time the master branch of the CMT repo
   (CMT:master) had been synced with nek5_svn.
2. We created two branches of the Nek5000 and CMT git repos
   (Nek5000:cmt-merge-1 and CMT:cmt-sync, respectively) at the
   downstream commit identified in (1).
3. CMT:cmt-sync was merged into cmt-merge-1.  The latter was pushed to
   RonRahaman/Nek5000.
4. A new branch (cmt-rebase-1) was made at the most recent commit of
   CMT:master.
5. cmt-rebase-1 was rebased on top of cmt-merge-1.  The former was
   pushed to RonRahaman/Nek5000.
In this commit,
1. We created a new branch (cmt-merge-2) at the most recent commit of
   Nek5000:develop.
2. We merged cmt-rebase-1 on top of cmt-merge-2
The only issues we needed to manually resolve were:
1. README.md was kept entirely from cmt-rebase-1 (i.e., from
   CMT:master)
2. tools/* were kept entirely from cmt-merge-2 (i.e., from
   Nek5000:develop)
3. tools/scripts/ was deleted; it had been copied from CMT:master in the
   merge, but was not needed
TODO: Check that these files were merged as intended:
1. core/ic.f
2. core/makefile.template
3. core/makenek.inc

RonRahaman committed:
Patching makenek.inc after previous merge

RonRahaman committed:
When IFCMT=fase, don't append cmt_switch to case_name.f

RonRahaman committed:
Moved CMT README to core/cmt. Restored original README

RonRahaman committed:
Merge pull request #1 from RonRahaman/cmt-merge-2

Moved CMT README to core/cmt. Restored original README

RonRahaman commented:
Right now, the CMT example problems are in cmt-cases/.  They should probably be moved elsewhere, either in short_tests/ or in the NekExamples repo.  

Our intent is to have regression tests for cmt-cases/ at some point in the near future.  Thoughts?

cc: @fischer1 @hackljf @stgeke

stgeke commented:
I suggest the following changes:
- move cmt out of core into ext/cmt (create an ext top-level directory)
- move cmt_cases into examples and/or short-tests 
- use a pre-processor flag IFCMT instead of logical flag in makenek
- remove all cmt specific code from core (we may need to do some changes to support extensions like this)

stgeke commented:
Why do we want to include something like CMT (which is experimental feature) into our stable release branch?

pschlatt1 commented:
I would agree on moving CMT out of core; but perhaps in the 3rd-party directory?

RonRahaman commented:
Those are great ideas, but moving cmt/ out of core/ should be a separate PR.  This PR contains bugfixes to the existing cmt/ code.  On the other hand, moving cmt/ is an enchancement.  PRs should represent atomic concepts.

stgeke commented:
Is this PR atomic? It looks more like an initial merge to me.

stgeke commented:
Can somebody remind me again about the intention of having a top-level 3rd-party directory?

pschlatt1 commented:
my understanding is exactly for module like cmt; non-essential add-ons predominantly developed, maintained and used by non-core developers.

stgeke commented:
Really? I don't think we should host external source code in our project.

pschlatt1 commented:
I guess it depends on the definition of external. We were thinking of e.g. Arnoldi which is non-essential, but still widely used and should be in-sync with the main development. But I might be wrong, of course.

RonRahaman commented:
It's not external.  @fischer1 is a co-PI of the CCMT at University of Florida.

stgeke commented:
Don't get me wrong - I'm not saying that CMT should be a separate project. I just thought it's a good opportunity to bring the discussion about extensions, plug-ins, toolboxes up again. This is something we need to sort out pretty soon.

hackljf commented:
Dear Ron, Stefan, Prof. Schlatter, and others,

I really appreciate your hard work getting CMT-nek integrated into nek5000
cleanly and consistently. I am recovering from surgery and busy with our DG
solver for Navier-Stokes and artificial viscosity as well as bringing some
old examples back up to date.

CMT-nek in its current form was designed as an "execution branch" from an
otherwise conventional run of nek5000. Paul and my former coworker, Mrugesh
Shringarpure, worked out the current method. The only changes to core/ were
meant to be IFCMT in the advance loop and nekuic. Unfortunately, restarting
required the odd situation where we needed the T field, but couldn't run
with the dssums done IFHEAT=T. Paul committed some workarounds to that. I
apologize that it got a little messier. I defer to whatever design
decisions you guys make. Just please teach us in Florida how to run with
them.

I will get you at least one example for short_cases today. If you change
makenek, please teach me what to do.

## 

Jason Hackl

On Fri, Sep 2, 2016 at 11:15 AM, Ron Rahaman notifications@github.com
wrote:

> Right now, the CMT example problems are in cmt-cases/. They should
> probably be moved elsewhere, either in short_tests/ or in the NekExamples
> repo.
> 
> Our intent is to have regression tests for cmt-cases/ at some point in the
> near future. Thoughts?
> 
> cc: @fischer1 https://github.com/fischer1 @hackljf
> https://github.com/hackljf @stgeke https://github.com/stgeke
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/Nek5000/Nek5000/pull/80#issuecomment-244403927, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AUG6zc5MVl8d7PAhAnpDJE-5ARnizNJjks5qmD2KgaJpZM4JzyOd
> .

RonRahaman commented:
@stgeke, I agree that the discussion about 3rd party modules is a good one to have.  

I was under the impression that the CMT improvements need to make their way into Nek5000 one way or another, and that's my only concern.  

I will of course defer to your judgement regarding:
- Whether the CMT bugfixes should be included in v16.0.0
- Whether CMT should live in 3rd_party/ eventually

Once you've decided what to do, I'd appreciate it if you raised an Issue outlining exactly what we need to do to get the CMT changes back into Nek.

Thanks!

stgeke commented:
I think this PR should be merged into `develop` and CMT (which is in an experimental stage) should be deleted from the `release` branch, at least for now.

hackljf commented:
I wouldn't worry about CMT bug fixes in 16.0.0. As long as the world at
large can download the homentropic vortex case from short_cases and run it
for at least a little while, that's fine. I am OK with removing CMT from
release and leaving it in develop AS LONG AS the develop branch is publicly
available to students at UF and I can issue pull requests to it.

I'm not sure what the intended spirit of 3rd_party is. It would be nice for
Paul to decide whether or not CMT-nek lives there.

## 

Jason Hackl

On Mon, Sep 5, 2016 at 10:50 AM, Stefan Kerkemeier <notifications@github.com

> wrote:
> 
> I think this PR should be merged into develop and CMT (which is in an
> experimental stage) should be deleted from the release branch, at least
> for now.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/Nek5000/Nek5000/pull/80#issuecomment-244764131, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AUG6zRyd5VaJ0pM_v7eTU2oJwGWu7kGOks5qnCwdgaJpZM4JzyOd
> .

RonRahaman commented:
Yes, @stgeke, I respect your opinion and you have some very valid points, but the decision to remove CMT should not be taken lightly.  @fischer1 had originally put CMT into Nek, so it should be his decision.  

IMO, it doesn't make sense to remove CMT from 'release/v16.0.0' just to add it back into 'develop' post-release.

stgeke commented:
Yes, I am not the one who decides. Keep in mind: In the past we had a development but no release branch. That's why we have some experimental things (not just cmt) in the release today. IMO, something we need to clean-up.

RonRahaman commented:
If you'd like to start removing features, I'd appreciate it if you stated your intentions in detail on the [Issues page](https://github.com/Nek5000/Nek5000/issues).  If you have a clear plan, I'm happy to support you.

RonRahaman commented:
> I suggest the following changes:
> 
> move cmt out of core into ext/cmt (create an ext top-level directory)
> move cmt_cases into examples and/or short-tests
> use a pre-processor flag IFCMT instead of logical flag in makenek
> remove all cmt specific code from core (we may need to do some changes to support extensions like this)

Right now, I'm happy to do the following.  They're unequivocably good ideas:
- [ ]  Move all cmt_cases into NekExamples
- [ ]  Add the CMT verificaiton to short_cases, Travis, and buildbot.  @hackljf just shared these with me
- [ ]  Replace IFCMT in makenek with a preprocessor flag in makenek.inc

I'm only comfortable doing the other action items (esp. deleting this feature from 'release/v16.0.0') after input from @fischer1.  @stgeke and @hackljf, is this satisfactory?

hackljf committed:
JH091416 First working code with Baumann-Oden DGSEM for higher-order
terms. Will eventually become compressible Navier-Stokes solver, and
host artificial viscosity for shocks. Developer docs will reflect this commit
and date.

hackljf committed:
More bugfixes. Somehow CMTDATA crept back into core/, but it's gone
again.

hackljf committed:
SEND IT ALL UPSTREAM THIS IS A HUGE BUGFIX IN EQNSOLVER_CMT
Turns out that evaluate_dealiased_conv_h was a huge mess; I had
written a needlessly quadratic (but pleasantly compact) flux
function for mass AND momentum, but only did about a third of the
job. Mass conservation was a mess. Back to setting interpolating
U onto H for the mass equation. Whew.

hackljf committed:
Merge pull request #2 from hackljf/cmt-merge-2

Cmt merge 2

hackljf committed:
turned most stuff on. integer*8 bug in dg_setup slipped through the cracks

hackljf committed:
MMerge branch 'cmtvisc' into cmt-merge-2

hackljf committed:
now with diffusion number. nav stokes + vortex blows up pretty quick.
now the fun really begins

hackljf committed:
name tweaks. nav stokes no worky. off to monolithic regularization for now

hackljf committed:
Viscous term development plodding along

hackljf committed:
Merge branch 'cmtvisc' into cmt-merge-2
BIG COMMIT AND BIG MERGE WITH THE MOTHERSHIP

hackljf committed:
Merge pull request #4 from hackljf/cmt-merge-2

Cmt merge 2

hackljf committed:
Monolithic regularization introduced, but no worky yet

hackljf committed:
bugfix, more debugging

RonRahaman committed:
Updated NekTests, .travis.yml to include CMT tests

RonRahaman committed:
Forgot to add short_tests/CMT/inviscid_vortex directory

RonRahaman committed:
Renamed file for reference solutions

hackljf committed:
Merge pull request #5 from RonRahaman/cmt-merge-2

Travis tests w/ inviscid_vortex

RonRahaman committed:
Removed cmt_cases directory

hackljf committed:
Merge pull request #7 from RonRahaman/cmt-merge-2

Removed cmt_cases directory

hackljf committed:
MONOLITHIC ARTVISC FINALLY WORKS;EVM OR BUST!!!!!

hackljf committed:
Merge branch 'cmtvisc' into cmt-merge-2

RonRahaman commented:
@stgeke, do you have a preference for the preprocessor symbol we use in PPLIST?  The symbols "CMT" and "IFCMT" are already in use.

hackljf committed:
Merge pull request #8 from hackljf/cmt-merge-2

Cmt merge 2

fischer1 commented:
I envision one or more compressible variants, perhaps to be merged into a single option down the road.   Compressible is something for which there is significant demand and will ultimately be on an equal footing with, say, low Mach as an option.  

I very much like the idea of supporting Arnoldi as well.  The one difficulty I have with adjoints and Arnoldi is that there seems to be a large case-to-case dependency (correct me if I'm wrong), which makes it difficult to live in the source, since one needs to modify things from one problem to the next.  (At least, that was my impression with adjoints --- and I merely hazard a guess that the same carries over for Arnoldi.)

My inclination is that, if it has broad use, is highly repeated code that would otherwise be duplicated in 100s of .usr files, then it probably belongs in "core".   The idea of the CMT subdirectory was simply to give the CMT group room to work, with a lightweight connection to the main source.   It's important that their code be in the Nek5000 repo.  I don't know that we need to put it in 3rd party.   3rd party, in my view, would be for something that is really distinct from Nek (e.g., moab, lapack) as opposed to something that is closely integrated with Nek.

I _think_ Arnoldi would be in the spirit of many of the post-processing / ancillary routines that are in postpro and navier5.f.  If it has a relatively stable interface, then it should be in core.

RonRahaman committed:
Now using line #2 of l2norms.dat.ref as reference values

hackljf committed:
Merge pull request #9 from RonRahaman/cmt-merge-2

Running inviscid_vortex test with 1000 timesteps

fischer1 commented:
@hackljf  Jason - can you resolve the conflicts for this merge request?  Paul

RonRahaman committed:
Merge branch 'release/v16.0.0' of github.com:Nek5000/Nek5000 into cmt-merge-2

stgeke commented:
I think of CMT as an extension which should be encapsulated with a well defined interface. To accept the pull request I recommend the following changes:

0) Pull in the latest changes and resolve the conflicts
1) Eliminate IFCMT in makenek and add a preprocessor symbol (CMT) instead
2) Compile source code in core/CMT if CMT is defined
3) Use the existing t-array for fields like (phig, …), no need to define extra arrays for them
4) Remove all CMT stuff from core except the top-level driver call in nek_advance() and ifcmt in INPUT
    - With (3) you don’t need to modify nekuic() anymore
    - Instead of putting your data into NEKUSE have your own CMTUSE which you include in the .usr file
    - Not sure how you could avoid the changes in nekasgn(), but may be you have a good idea?

We can consider this pull request for v16.0.0 if you can implement the suggested changes within the next 14 days.

hackljf commented:
Thank you so much for your management and patient suggestions. With Ron's
or somebody else's help introducing a preprocessor symbol for step 1, I
will have all of these done in 14 days. I think I can also write a
dedicated nekasgn for CMTUSE and leave it somewhere in the cmt/ directory's
source code.

What is the maximum value of ldimt I can use?

## 

Jason Hackl

On Thu, Nov 10, 2016 at 12:54 PM, Stefan K. notifications@github.com
wrote:

> I think of CMT as an extension which should be encapsulated with a well
> defined interface. To accept the pull request I recommend the following
> changes:
> 
> 0) Pull in the latest changes and resolve the conflicts
> 1) Eliminate IFCMT in makenek and add a preprocessor symbol (CMT) instead
> 2) Compile source code in core/CMT if CMT is defined
> 3) Use the existing t-array for fields like (phig, …), no need to define
> extra arrays for them
> 4) Remove all CMT stuff from core except the top-level driver call in
> nek_advance() and ifcmt in INPUT
> - With (3) you don’t need to modify nekuic() anymore
> - Instead of putting your data into NEKUSE have your own CMTUSE which you
>   include in the .usr file
> - Not sure how you could avoid the changes in nekasgn(), but may be you
>   have a good idea?
> 
> We can consider this pull request for v16.0.0 if you can implement the
> suggested changes within the next 14 days.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/Nek5000/Nek5000/pull/80#issuecomment-259760000, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AUG6zTikh_ESTJGuCo4h92DC5Y0PE3XDks5q81pGgaJpZM4JzyOd
> .

stgeke commented:
Forgot to add: You should also eliminate the CMT stuff in SIZE. I don't think this is needed anymore with the preprocessor flag and the t-array.

It's up to you (the user) to set a value for ldimt in SIZE.

stgeke commented:
I can help you with the preprocessor stuff and the build system. Just go ahead and implement the suggest changes and I'll do the rest.

RonRahaman committed:
Cleaning up some mistakes from merge

RonRahaman committed:
Fixes to short_tests/lib/nekTestCase.py

This commit brings nekTestCase.py back in line with
Nek5000:release/v16.0.0.  However, it has still diverged from
NekExamples:master.  This should be addressed in the future.

hackljf commented:
right, but I'm asking if, in classic nek, ldimt is a number that is
typically less than 7 or not. We need at least this many variables, and
there's a good chance we'll need more.

On Thu, Nov 10, 2016 at 1:46 PM, Stefan K. notifications@github.com wrote:

> Forgot to add: You should also eliminate the CMT stuff in SIZE. I don't
> think this is needed anymore with the preprocessor flag and the t-array.
> 
> It's up to you (the user) to set a value for ldimt in SIZE.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/Nek5000/Nek5000/pull/80#issuecomment-259772847, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AUG6zXBfJ2xTqOSOhleX4Y4YsvInu-WMks5q82ZwgaJpZM4JzyOd
> .

hackljf commented:
thank you for your help.

On Thu, Nov 10, 2016 at 3:59 PM, Jason Hackl hackljf@gmail.com wrote:

> right, but I'm asking if, in classic nek, ldimt is a number that is
> typically less than 7 or not. We need at least this many variables, and
> there's a good chance we'll need more.
> 
> On Thu, Nov 10, 2016 at 1:46 PM, Stefan K. notifications@github.com
> wrote:
> 
> > Forgot to add: You should also eliminate the CMT stuff in SIZE. I don't
> > think this is needed anymore with the preprocessor flag and the t-array.
> > 
> > It's up to you (the user) to set a value for ldimt in SIZE.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > https://github.com/Nek5000/Nek5000/pull/80#issuecomment-259772847, or mute
> > the thread
> > https://github.com/notifications/unsubscribe-auth/AUG6zXBfJ2xTqOSOhleX4Y4YsvInu-WMks5q82ZwgaJpZM4JzyOd
> > .

stgeke commented:
The limit is and was always the available memory.

RonRahaman committed:
Fixed CMT/inviscid_vortex test case

hackljf committed:
Merge pull request #12 from RonRahaman/cmt-merge-2

Merging (w/conflicts) from upstream release/v16.0.0

RonRahaman commented:
This branch is back in sync with the upstream changes and passes all the tests.  Can we merge it now and work on the IFDEFs as a separate Issue?  That way, someone else can claim the issue and decouple it from these changes.  

cc: @stgeke @hackljf @fischer1

hackljf commented:
That gets my vote.

On Fri, Nov 11, 2016 at 10:01 AM, Ron Rahaman notifications@github.com
wrote:

> This branch is back in sync with the upstream changes and passes all the
> tests. Can we merge it now and work on the IFDEFs as a separate Issue? That
> way, someone else can claim the issue and decouple it from these changes.
> 
> cc: @stgeke https://github.com/stgeke @hackljf
> https://github.com/hackljf @fischer1 https://github.com/fischer1
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/Nek5000/Nek5000/pull/80#issuecomment-259975509, or mute
> the thread
> https://github.com/notifications/unsubscribe-auth/AUG6zfWLi1gmWoR4lZdAgv8hrFn-pPC_ks5q9INbgaJpZM4JzyOd
> .

stgeke commented:
This PR contains precore and postcore in tools. What's that?

hackljf committed:
GRAND EFFORT TO GET CMT OUT OF NON-CMT COMPILATION. COMPARE IC

hackljf committed:
Ready to remove IFCMT and transition to compiler directive

hackljf committed:
NOW THIS IS THE CORRECT USR FILE I DO NOT EVER WANT TO SEE VARSIC DECLARED IN L2NORM EVER AGAIN

hackljf committed:
python fix for which line to read from l2norm.dat.ref

hackljf committed:
Merge pull request #14 from hackljf/cmt-merge-2

halfway to v16.0.0.0 from this end. onward to conditional compilation

hackljf committed:
Broke makenek pretty hard, but I'm committing anyway because nek5000 still works fine

hackljf committed:
code works with and without -DCMTNEK. fix .travis.yml

hackljf committed:
forgot to update pvort.rea. removed IFFLTR from pvort.rea

hackljf committed:
I'm going to try this: replace IFCMT with PPLIST and see if that gets exported

hackljf committed:
Re-inserted IFCMT in .travis.yml. Will PPLIST do what I mean?

hackljf committed:
Merge pull request #16 from hackljf/cmt-merge-2

I believe I have satisfied the requirements laid forth by the nek community to bring CMT-nek into v16.0.0.0. Compiling nek5000 now proceeds without any cmt-nek code at all; it all lives in core/cmt. IFCMT has been replaced by CMTNEK, a compiler directive that appears in three places in two source files. travis and NekTests.py reflect this accurately. Let me know if I need to do anything else. Thank you for your patience.

stgeke commented:
You did a good job!

Please do the following minor corrections:
- prefix the routines you add to the .usr file (e.g. usrflt -> cmt_usrflt) to avoid name conflicts
- remove core/README
- remove tools/precore
- remove tools/postcore
- remove short_tests/CMT/inviscid_vortex/fort.144
- use lastest SIZE.template for short_tests/CMT/SIZE

hackljf committed:
some tweaks for v16.0.0.0

hackljf committed:
Merge pull request #17 from hackljf/cmt-merge-2

some tweaks for v16.0.0.0

RonRahaman commented:
@hackljf, I'm making some fixes so that PPLIST will work in NekTests as intended.  

Some necessary changes to short_tests/lib were just merged in #158.  They introduced some expected conflicts in this PR that I will resolve.  

I also need to make a few changes to the CmtInviscidVortex class in NekTests.py.  Stay tuned for a PR in ufccmt/Nek5000

RonRahaman committed:
Merge remote-tracking branch 'upstream/release/v16.0.0' into cmt-merge-2

# Conflicts:
#	.travis.yml
#	short_tests/lib/nekBinBuild.py
#	short_tests/lib/nekFileConfig.py
#	short_tests/lib/nekTestCase.py

RonRahaman committed:
Merge remote-tracking branch 'upstream/release/v16.0.0' into cmt-merge-2

# Conflicts:
#	.travis.yml
#	short_tests/lib/nekTestCase.py

hackljf committed:
see if this works. deleted a filter common block. the first of many

RonRahaman committed:
Fixed interfaces for CmtInviscidVortex test case

hackljf committed:
Merge pull request #18 from hackljf/cmt-merge-2

see if this works. deleted a filter common block. the first of many

hackljf committed:
Merge pull request #19 from RonRahaman/cmt-merge-2

Updated CMT regression tests

hackljf committed:
Merge pull request #1 from UFCCMT/cmt-merge-2

Cmt merge 2

RonRahaman commented:
To update everyone, NekTests.py has been updated to support the new compilation options for CMTNEK.  

Previously, NekTests would set IFCMT=true in makenek to run the CmtInviscidVortex test case.  Now, NekTests will set CMTNEK in PPLIST to run that test case.  

These changes were merged into UFCCMT/Nek5000:cmt-merge-2 and automagically updated in this PR.

stgeke commented:
The cmt/README.md looks pretty specific to me. Is this file really needed?
Is there anything else we need to do before we can merge this PR?

hackljf committed:
README.md removed

hackljf committed:
Merge pull request #20 from hackljf/cmt-merge-2

Cmt merge 2

hackljf commented:
Removed and merged

On Wed, Nov 23, 2016 at 4:11 PM, Stefan K. <notifications@github.com> wrote:

> The cmt/README.md looks pretty specific to me. Is this file really needed?
> Is there anything else we need to do before we can merge this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Nek5000/Nek5000/pull/80#issuecomment-262630843>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AUG6zfpDfTugFT7QMSVa9X-JERUqvQehks5rBKv5gaJpZM4JzyOd>
> .
>

stgeke commented:
@fischer1 I recommend to merge this PR after the conflict was resolved.

stgeke commented:
@hackljf Please resolve the conflict.

hackljf committed:
trying to resolve conflict in ic.f that doesn't involve code I ever use

hackljf commented:
I don't understand conflicts yet at all. When I tried to emulate some of
the changes in v16.0.0.0's ic.f that were not in cmt-merge-2's ic.f. it
broke PnPn (but everything else passed UFCCMT's regression testing).

Whatever the version of ic.f I am conflicting with looks like (and I can't
tell if it's the one I see in release/v16.0.0.0 or not), I have fulifilled
my promise of removing ALL cmt code and declarations from it with the
UNIQUE exception of

 #ifdef CMTNEK
         ifgett=.true. ! CMT-nek still not compatible with IFHEAT
#endif

near the end of sioflag. Otherwise, the characters "c," "m" and "t" do not
appear near each other because of CMT-nek.

 I am deeply sorry I don't know what else I can do.

-- 

Jason Hackl

On Thu, Nov 24, 2016 at 9:14 AM, Stefan K. <notifications@github.com> wrote:

> @hackljf <https://github.com/hackljf> Please resolve the conflict.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Nek5000/Nek5000/pull/80#issuecomment-262785790>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AUG6zXwKJtVgG_bWUX5t-kd6yOumfUYSks5rBZvcgaJpZM4JzyOd>
> .
>

RonRahaman commented on core/ic.f:
```diff
@@ -56,14 +50,6 @@ subroutine setics
 
       jp = 0                  ! Set counter for perturbation analysis
 
-      if (ifcmt) then
```
The only conflicts between this branch (cmt-merge-2) and release/v16.0.0 occurs here.  The diff looks like this, where the upper referes to the changes in cmt-merge-2 and the lower refers to release/v16.0.0:

```diff
++<<<<<<< cmt-merge-2
 +      irst = param(46)        ! for lee's restart (rarely used)
 +      if (irst.gt.0)  call setup_convect(2)
 +
++=======
+       if (ifcmt) then
+          call rzero(phig,ltott)
+          call rzero(csound,ltott)
+          call rzero(vtrans,ltott*ldimt1)
+          call rzero(vdiff ,ltott*ldimt1)
+          call rzero(u,ntotcv)
+       endif
++>>>>>>> release/v16.0.0
```

RonRahaman reviewed (commented):
@hackljf, you resolve the merge on the command line like this.  Suppose you have the following remotes defined:
* 'origin' points to https://github.com/hackljf/Nek5000.git 
* 'upstream' points to https://github.com/Nek5000/Nek5000.git 
* 'ufccmt' points to https://github.com/UFCCMT/Nek5000.git.  

Then you do the following:
```
git checkout cmt-merge-2
git pull ufccmt cmt-merge-2    # Make sure your local branch is up-to-date
git pull upstream release/v16.0.0    # Uh-oh!  Conflicts!
```

As you noted, the conflicts are in core/ic.f.  The conflicts are limited to the lines noted here (in the conversation for this review).  Luckily, these are small conflicts and you can just clean it up in your text editor (no need to use a fancy merge tool).  Once you've cleaned-up the file to your liking, you do the following

```
git add core/ic.f    # Mark the conflicts as resolved
git commit    # Commit the merge
git push origin cmt-merge-2    # Push your changes to your fork
```
Then you submit following PR:
https://github.com/UFCCMT/Nek5000/compare/cmt-merge-2...hackljf:cmt-merge-2

Once the above PR is merged, then this PR (the one we're discussing) will be updated and shouldn't have any conflicts.

hackljf commented:
Thank you. I will tend to this and follow your instructions in the next
three hours

On Friday, November 25, 2016, Ron Rahaman <notifications@github.com> wrote:

> *@RonRahaman* commented on this pull request.
>
> @hackljf <https://github.com/hackljf>, you resolve the merge on the
> command line like this. Suppose you have the following remotes defined:
>
>    - 'origin' points to https://github.com/hackljf/Nek5000.git
>    - 'upstream' points to https://github.com/Nek5000/Nek5000.git
>    - 'ufccmt' points to https://github.com/UFCCMT/Nek5000.git. Then you
>    do the following:
>
> git checkout cmt-merge-2
> git pull ufccmt cmt-merge-2  # Make sure your local branch is up-to-date
> git pull upstream release/v16.0.0  # Uh-oh!  Conflicts!
>
> As you noted, the conflicts are in v16.0.0. The conflicts are limited to
> the lines noted here. Luckily, these are small conflicts and you can just
> clean it up manually in your text editor (no need to use a fancy merge
> tool). Once you've edited the file to your liking, you do the following
>
> git add core/ic.f  # Mark the conflicts as resolved
> git commit # Commit the merge
> git push origin cmt-merge-2 # Push your changes to your fork
>
> Then you submit following PR:
> UFCCMT@cmt-merge-2...hackljf:cmt-merge-2
> <https://github.com/UFCCMT/Nek5000/compare/cmt-merge-2...hackljf:cmt-merge-2>
>
> Once the above PR is merged, then this PR (the one we're discussing) will
> be updated and shouldn't have any conflicts.
> ------------------------------
>
> In core/ic.f
> <https://github.com/Nek5000/Nek5000/pull/80#pullrequestreview-10181265>:
>
> > @@ -56,14 +50,6 @@ subroutine setics
>
>        jp = 0                  ! Set counter for perturbation analysis
>
> -      if (ifcmt) then
>
> The only conflicts between this branch (cmt-merge-2) and release/v16.0.0
> occurs here. The diff looks like this, where the upper referes to the
> changes in cmt-merge-2 and the lower refers to release/v16.0.0:
>
> ++<<<<<<< cmt-merge-2
>  +      irst = param(46)        ! for lee's restart (rarely used)
>  +      if (irst.gt.0)  call setup_convect(2)
>  +++=======+       if (ifcmt) then+          call rzero(phig,ltott)+          call rzero(csound,ltott)+          call rzero(vtrans,ltott*ldimt1)+          call rzero(vdiff ,ltott*ldimt1)+          call rzero(u,ntotcv)+       endif++>>>>>>> release/v16.0.0
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Nek5000/Nek5000/pull/80#pullrequestreview-10181265>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AUG6zRTZ8xmH6_evDVkhW7ZqIqnBGpFpks5rBwemgaJpZM4JzyOd>
> .
>

hackljf committed:
Revert "trying to resolve conflict in ic.f that doesn't involve code I ever use"

This reverts commit c219fa1afcb73db107cc36304a85863af8fcc6d6.

hackljf committed:
Merge pull request #22 from hackljf/cmt-merge-2

undoing failed conflict resolution in core/ic.f

hackljf committed:
Merge pull request #2 from UFCCMT/cmt-merge-2

Merge pull request #22 from hackljf/cmt-merge-2

hackljf committed:
Merge pull request #23 from hackljf/cmt-merge-2

pinball

hackljf committed:
Merge pull request #3 from UFCCMT/cmt-merge-2

Merge pull request #23 from hackljf/cmt-merge-2

hackljf committed:
Merge branch 'release/v16.0.0' of https://github.com/Nek5000/Nek5000 into cmt-merge-2

Conflicts:
	core/ic.f

hackljf committed:
maybe this will save the day re: travis

hackljf committed:
I stand by my code, but someting happened during the last conflict resolve.
I will not have time to deal with it  until after the CCMT tri-lab
services Team review on December 2. CMT tests skipped for now. commits before
11/23/2016 are trustworthy to CMT-nek.

hackljf committed:
Merge pull request #24 from hackljf/cmt-merge-2

fixed ic.f conflicts, but CMT/invscid_vortex now fails subtly. I have accepted defeat and turned off this test in .travis.yml. I trust my code, but I will not have time to decipher what, exactly "Lee's restart" and extra setup_convect(2) that I am sure I skipped have done to make my exemplary example stop working until after December 2
